auth/src/main/java/io/grpc/auth/ClientAuthInterceptor.java

```diff
@@ -39,40 +40,69 @@
 import io.grpc.ClientInterceptors.ForwardingCall;
 import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
+import io.grpc.Status;
 
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
 import java.util.concurrent.Executor;
 
-import javax.inject.Provider;
+/**
+ * Client interceptor that authenticates all calls by binding header data provided by a credential.
+ * Typically this will populate the Authorization header but other headers may also be filled out.
+ *
+ * <p> Uses the new and simplified Google auth library:
+ * https://github.com/google/google-auth-library-java
+ */
+public class ClientAuthInterceptor implements ClientInterceptor {
 
-/** Client interceptor that authenticates all calls with OAuth2. */
-public class OAuth2ChannelInterceptor implements ClientInterceptor {
-  private static final Metadata.Key<String> AUTHORIZATION =
-      Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
+  private final Credentials credentials;
 
-  private final OAuth2AccessTokenProvider accessTokenProvider;
-  private final Provider<String> authorizationHeaderProvider
-      = new Provider<String>() {
-        @Override
-        public String get() {
-          return "Bearer " + accessTokenProvider.get();
-        }
-      };
+  private Metadata.Headers cached;
+  private Map<String, List<String>> lastMetadata;
 
-  public OAuth2ChannelInterceptor(Credential credential, Executor executor) {
-    this.accessTokenProvider = new OAuth2AccessTokenProvider(credential, executor);
+  public ClientAuthInterceptor(Credentials credentials, Executor executor) {
+    this.credentials = Preconditions.checkNotNull(credentials);
   }
 
   @Override
   public <ReqT, RespT> Call<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
-      Channel next) {
+                                                       Channel next) {
     // TODO(ejona86): If the call fails for Auth reasons, this does not properly propagate info that
     // would be in WWW-Authenticate, because it does not yet have access to the header.
     return new ForwardingCall<ReqT, RespT>(next.newCall(method)) {
       @Override
       public void start(Listener<RespT> responseListener, Metadata.Headers headers) {
-        headers.put(AUTHORIZATION, authorizationHeaderProvider.get());
-        super.start(responseListener, headers);
+        try {
+          synchronized (this) {
+            // TODO(lryan): This is icky but the current auth library stores the same
+            // metadata map until the next refresh cycle. This will be fixed once
+            // https://github.com/google/google-auth-library-java/issues/3
+            // is resolved.
+            if (lastMetadata == null || lastMetadata != credentials.getRequestMetadata()) {
+              lastMetadata = credentials.getRequestMetadata();
```

Comment:
Its not expensive at all, whats relatively expensive is the conversion we have to do on it